sdk/attestation/azure-security-attestation/inc/azure/attestation/attestation_administration_client.hpp

```diff
@@ -0,0 +1,109 @@
+   * -# AAD Mode
+   * -# Isolated Mode
+   *
+   * Shared mode attestation service instances do not allow any administration actions at all. They
+   * exist to allow customers to perform attestation operations without requiring any
+   * customizations.
+   *
+   * AAD Mode instances allow customers to modify attestation policies. When the attestation
+   * instance is in AAD mode, the creator of the instance indicates that they trust ARM RBAC and
+   * Microsoft AAD to validate client connections to the service. As such, additional authentication
+   * is not required for administrative operations.
+   *
+   *
+   */
+
+  class AttestationAdministrationClient final {
+
+  public:
+    /**
+     * @brief Destructor.
+     *
+     */
+    virtual ~AttestationAdministrationClient() = default;
+
+    /**
+     * @brief Construct a new Attestation Administration Client object
+     *
+     * @param endpoint The URL address where the client will send the requests to.
+     * @param credential The authentication token to use.
+     * @param options The options to customize the client behavior.
+     */
+    explicit AttestationAdministrationClient(
+        std::string const& endpoint,
+        std::shared_ptr<Core::Credentials::TokenCredential const> credential,
+        AttestationAdministrationClientOptions const& options
+        = AttestationAdministrationClientOptions());
+
+    /**
+     * @brief Returns the API version the client was configured with.
+     *
+     * @returns The API version used when communicating with the attestation service.
+     */
+    std::string const& ClientVersion() const { return m_apiVersion; }
+
+    /**
+     * @brief Construct a new Attestation Administration Client object from another attestation
+     * administration client.
+     *
+     * @param attestationClient An existing attestation client.
+     */
+    explicit AttestationAdministrationClient(
+        AttestationAdministrationClient const& attestationClient)
+        : m_endpoint(attestationClient.m_endpoint), m_apiVersion(attestationClient.m_apiVersion),
+          m_pipeline(attestationClient.m_pipeline),
+          m_tokenValidationOptions(attestationClient.m_tokenValidationOptions){};
+
+    /** @brief Retrieves the attestation policy for the specified attestation type.
+     */
+
+    Response<Models::AttestationToken<std::string>> GetAttestationPolicy(
+        Models::AttestationType const& attestationType,
+        GetPolicyOptions const& options = GetPolicyOptions(),
+        Azure::Core::Context const& context = Azure::Core::Context::ApplicationContext) const;
+
+  private:
+    Azure::Core::Url m_endpoint;
+    std::string m_apiVersion;
+    std::shared_ptr<Azure::Core::Credentials::TokenCredential const> m_credentials;
+    std::shared_ptr<Azure::Core::Http::_internal::HttpPipeline> m_pipeline;
+    AttestationTokenValidationOptions m_tokenValidationOptions;
+
+    mutable std::vector<Models::AttestationSigner> m_attestationSigners;
```

Comment:
I haven't seen this `mutable` keyword being used before in our SDK. For my edification, what's the purpose of it, generally, and why do we need to use it in this instance?

Reply:
We claim that SDK clients are immutable. That means that service methods should all be declared as "const". 

By using `mutable` we limit the potential damage of breaking const-ness to just the variables which are known to be mutable.

Reply:
Again, up to you, but it feels wrong for us to claim one thing, convey intent using `const` and then break it.

If we want to limit the potential damage of breaking const-ness, rather than marking the variable as mutable, why not mark the method that is mutating it as non-const? And now the intent is clear? Or is that less desirable in this case compared to using `mutable`?

Without that, this variable can be mutated in all `const` methods, rather than isolating it to just the few that need to mutate it. Unless, all of them do mutate it, then that's a different question :)
Is that the case here?

Maybe @JeffreyRichter has some thoughts on immutability of SDK clients and when we should use `mutable` vs `const` on the particular method itself.

Reply:
The compiler will not allow you to call a non-const method from a const method. The only way I know of working around that is the `const_cast<Foo*>(this)` trick, and that 100% *is* code smell.

Reply:
The signing certificates are per-attestation instance (per-url), not per-client. But if you have 5 clients talking to 5 separate attestation service instances you will have five separate sets of certificates.

So implementing an attestation service specific pipeline policy to handle caching these values might be possible. It would be tricky, because we only want to have it enabled on certain API calls (it would be bad to invoke this policy on the `GetAttestationSigningCertificates` API for example). Another slight complication: If the customer says they don't want token validation, we don't ever want to retrieve the signing certificates. And that happens on a per-API basis (note: Not currently implemented - but it's on the list).

But it is worth considering, since we would have complete control over the policy.

OTOH, All this does is to move the mutable item from the client into a policy object. It doesn't change the architecture.